vivarium/core/serialize.py

```diff
@@ -1,62 +1,39 @@
 
 import numpy as np
-from bson import ObjectId
 from pint import Unit
 from pint.quantity import Quantity
+from bson.codec_options import TypeEncoder, TypeRegistry, CodecOptions
+from bson import _dict_to_bson, _bson_to_dict
 
-from vivarium.core.registry import serializer_registry, Serializer
-from vivarium.core.composer import Composer, Composite
-from vivarium.core.process import Process
-from vivarium.library.dict_utils import remove_multi_update
 from vivarium.library.units import units
+from vivarium.core.registry import serializer_registry, Serializer
 
-
-# Serialization and deserialization functions that handle arbitrary data
-# types.
-
-
-def serialize_value(value: Any) -> Any:
-    # Try to lookup by exclusive type
-    serializer = serializer_registry.access(str(type(value)))
-    # If lookup fails, search through registered serializers
-    if not serializer:
-        compatible_serializers = []
-        for serializer_name in serializer_registry.list():
-            serializer = serializer_registry.access(serializer_name)
-            if serializer.can_serialize(value):
-                compatible_serializers.append(serializer)
-        if not compatible_serializers:
-            raise ValueError(
-                f'No serializer found for {value} of type {type(value)}')
-        if len(compatible_serializers) > 1:
-            raise ValueError(
-                f'Multiple serializers ({compatible_serializers}) found '
-                f'for {value} of type {type(value)}')
-        serializer = compatible_serializers[0]
-        if not isinstance(value, Process):
-            # We don't warn for processes because since their types
-            # based on their subclasses, it's not possible to avoid
-            # searching through the serializers.
-            warnings.warn(
-                f'Searched through serializers to find {serializer} '
-                f'for a value of type {type(value)}. This is '
-                f'inefficient.')
-    return serializer.serialize(value)
-
-
+def serialize_value(
+    value: Any,
+    codec_options: CodecOptions=None
+) -> Any:
+    if not codec_options:
+        codec_options = get_codec_options()
+    value = _dict_to_bson(value, False, codec_options)
+    return _bson_to_dict(value, codec_options)
+
+# Deserialization still requires custom python code because
+# BSON C extensions cannot distinguish between strings that
+# should be deserialized as different types (e.g. Units)
 def deserialize_value(value: Any) -> Any:
     compatible_serializers = []
     for serializer_name in serializer_registry.list():
         serializer = serializer_registry.access(serializer_name)
         if serializer.can_deserialize(value):
             compatible_serializers.append(serializer)
     if not compatible_serializers:
-        raise ValueError(
-            f'No deserializer found for {value}')
+        # Most likely a built-in type with no custom serializer/deserializer
```

Comment:
Unless the user makes an error and forgets to override the `deserialize` and `can_deserialize` methods when they override the `serialize` method, all values that enter this if block should simply be built-in Python types now that I've removed the `IdentitySerializer`. I removed that serializer because we can no longer assume a catch-all regex pattern to avoid using it on strings generated by custom serializers. I'll add in some kind of check to make sure all three serialization methods are overriden at the same time.

Reply:
I'm honestly not sure how to ensure that the user does not mess this up. I made the documentation for the serializer API (at the top of `registry.py`) very clear about what needs to be overriden in all possible cases. Hopefully that's enough to avoid any unexpected behavior from forgetting to override the correct methods.